Default Precompiled .NET templates to using track 2 extensions

hossam-nasr opened the pull request:
This PR defaults all precompiled .NET templates to using track 2 Azure SDK extensions, and updates the templates accordingly.

Affected templates:
* C# CosmosDB trigger
* F# CosmosDB trigger
* C# EventGrid trigger
* F# EventGrid trigger
* C# EventHub trigger
* F# EventHub trigger
* C# Storage Queue trigger
* F# Storage Queue trigger 
* C# SendGrid Trigger
* C# ServiceBus Queue trigger
* C# ServiceBus Topic trigger

hossam-nasr committed:
default queue trigger to track 2

default servicebus .NET isolated to track2

default eventhub to track2

update eventgrid

update sendgrid

default cosmosdb to track2

upgrade sendgrid extension to 3.0.2

update cosmosdb trigger

use normal logger in SB topic template

update SendGrid templates

update ServiceBusTopicTriggers

update ServiceBusTopic F#

add comment for CosmosDb templates

make sure cosmosdb values match templates

liliankasem commented:
@hossam-nasr hey Hossam, can you resolve conflicts here and I can help get this merged

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+      "args": {
+        "referenceType": "package",
+        "reference": "Microsoft.Azure.WebJobs.Extensions.CosmosDB",
+        "version": "4.0.0",
```
Can we use the latest version 4.3.0?

hossam-nasr replied:
Yes, this is a pretty old PR at this point, and the latest package at the time was 4.0.0. Updated to `4.3.0`

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+  "author": "Microsoft",
+  "classifications": ["Azure Function", "Trigger", "Cosmos DB"],
+  "name": "CosmosDBTrigger",
+  "identity": "Azure.Function.CSharp.CosmosDBTrigger.2.x",
```
Is this supposed to be 2.x or should it match the extension version 4.x? Also, should the folder name match the extension version 4.x instead of 5.x?

hossam-nasr replied:
For the first question, I will defer that to @soninaren. 

For the folder name, ideally, yes, but I named it `-5.x` because I saw that there were already `-5.x` folders for PowerShell, JavaScript, and Python. Would you know why those were named `-5.x`? Ultimately, though, folder names don't really matter that much, what matters is they are properly referenced in the ItemTemplates.nupkg file

liliankasem replied:
I'm surprised to see we have folders named 5.x; the value is supposed to represent the host version I think - at least that's the case for the project templates. And for track two, there's a new folder for track 2 templates called "Templates-v2". 

I'll sync with @soninaren and see what we want to do about the state of the repo folders + identity

hossam-nasr replied:
I believe that for item templates, the suffix is supposed to represent the version of the extension, which is `5.x` for most track 2 Azure SDK extensions. CosmosDB is one of few exceptions where their track 2 extension is version `4.x`. Also, I want to clarify that here, when we say track 2, we're talking about the versions of the extensions themselves. My understanding is that `Templates-v2` refers to v2 of the template schema, regardless of extension or host version.

liliankasem replied:
That makes sense, thanks for the extra context

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+      "replaces": "Company.Function",
+      "type": "parameter"
+    },
+    "ConnectionStringSetting": {
```
This is called "Connection" in this version of the extension

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+      "replaces": "ConnectionValue"
+    },
+    "DatabaseName": {
+      "description": "This is the name of the database that includes the collection to be monitored.",
```
collection --> container

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+      "defaultValue": "databaseName",
+      "replaces": "DatabaseValue"
+    },
+    "CollectionName": {
+      "description": "This is the name of the collection to be monitored.",
+      "type": "parameter",
+      "defaultValue": "collectionName",
+      "replaces": "CollectionValue"
```
Collection --> container

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/vs-2017.3.host.json:
```diff
@@ -0,0 +1,47 @@
+    "id": "2010"
+  },
+  "description": {
+    "text": "A C# function that will be run whenever documents change in a document collection.",
```
collection --> container

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/vs-2017.3.host.json:
```diff
@@ -0,0 +1,47 @@
+  ],
+  "symbolInfo": [
+      {
+        "id": "ConnectionStringSetting",
```
ConnectionStringSetting --> Connection

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-CSharp-5.x/.template.config/vs-2017.3.host.json:
```diff
@@ -0,0 +1,47 @@
+        "isVisible": true
+      },
+      {
+        "id": "CollectionName",
+        "name": {
+          "id": "2013",
+          "text": "Collection name",
```
Collection --> container

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-FSharp-4.x/.template.config/template.json:
```diff
@@ -0,0 +1,78 @@
+      "replaces": "Company.Function",
+      "type": "parameter"
+    },
+    "ConnectionStringSetting": {
```
Same naming comments as the CSharp version of the template, ConnectionStringSetting --> Connection and Collection --> Container

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-FSharp-4.x/.template.config/vs-2017.3.host.json:
```diff
@@ -0,0 +1,47 @@
+  ],
+  "symbolInfo": [
+      {
+        "id": "ConnectionStringSetting",
```
Naming consistency

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-FSharp-4.x/CosmosDBTriggerFSharp.fs:
```diff
@@ -0,0 +1,14 @@
+
+module CosmosDBTriggerFSharp =
+    [<FunctionName("CosmosDBTriggerFSharp")>]
+    let run([<CosmosDBTrigger(databaseName="DatabaseValue", collectionName="CollectionValue", ConnectionStringSetting="ConnectionValue", LeaseCollectionName="leases")>] input: IReadOnlyList<Document>, log: ILogger) =
```
Naming consistency

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-FSharp-4.x/function.json:
```diff
@@ -0,0 +1,14 @@
+        "type": "cosmosDBTrigger",
+        "name": "input",
+        "direction": "in",
+        "connectionStringSetting": "",
+        "databaseName": "",
+        "collectionName": "",
+        "leaseCollectionName": "leases",
+        "createLeaseCollectionIfNotExists" : true
```
Naming consistency

jcocchi commented on Functions.Templates/Templates/CosmosDBTrigger-FSharp-4.x/metadata.json:
```diff
@@ -0,0 +1,19 @@
+    "categoryStyle": "cosmosDB",
+    "enabledInTryMode": false,
+    "userPrompt": [
+      "connectionStringSetting",
+      "databaseName",
+      "collectionName",
+      "leaseCollectionName",
+      "createLeaseCollectionIfNotExists"
```
Naming consistency

jcocchi commented on Build/PackageFiles/Dotnet_precompiled/ItemTemplates_v3.x.nuspec:
```diff
@@ -25,17 +25,18 @@
+    <file src="..\..\..\Functions.Templates\Templates\EventGridTrigger-FSharp-3.x\**" target="content\EventGridTrigger-FSharp" exclude="**\*.csx;**\*.md;**\*.dat;**\function.json;**\metadata.json" />
     <file src="..\..\..\Functions.Templates\Templates\IoTHubTrigger-CSharp\**" target="content\IotHubTrigger-CSharpVS" exclude="**\*.csx;**\*.md;**\*.dat;**\function.json;**\metadata.json" />
-    <file src="..\..\..\Functions.Templates\Templates\CosmosDBTrigger-CSharp\**" target="content\CosmosDBTrigger-CSharp" exclude="**\*.csx;**\*.md;**\*.dat;**\function.json;**\metadata.json" />
+    <file src="..\..\..\Functions.Templates\Templates\CosmosDBTrigger-CSharp-5.x\**" target="content\CosmosDBTrigger-CSharp" exclude="**\*.csx;**\*.md;**\*.dat;**\function.json;**\metadata.json" />
```
Should this be 4.x to align with the current version of the Cosmos extension?

hossam-nasr replied:
The value here needs to match the folder name where the templates reside. See my comment above for why I named the folder `-5.x`, but I could easily rename it to `-4.x`

jcocchi reviewed (commented):
Some of the Cosmos DB names don't match the expected values for the 4.x extension. I have a PR specifically for Cosmos [here](https://github.com/Azure/azure-functions-templates/pull/1401), maybe we can split the Cosmos pieces out of this one so there aren't conflicts?

lilyjma commented:
> @hossam-nasr hey Hossam, can you resolve conflicts here and I can help get this merged

@hossam-nasr any update on this? 
Also, regarding Justine's comments, she made another PR specifically for Cosmos that has all the things she commented above (other than the project references that defaults to using Track 2): https://github.com/Azure/azure-functions-templates/pull/1401. I think the best approach here would be to 1. address the conflicts Lilian mentioned 2. work with @soninaren to see how to deal with the PR overlap (maybe we should remove the overlap and use her PR as the "source of truth" for Cosmos)

hossam-nasr committed:
fix eventgrid templates

hossam-nasr committed:
update package versions

hossam-nasr committed:
update cosmosdb settings names

hossam-nasr commented:
Hi @lilyjma / @liliankasem / @jcocchi , I've fixed all the merge conflicts in this PR and updated the templates to the latest versions of the extensions. This should be ready for review now. Thanks!

liliankasem commented:
> Hi @lilyjma / @liliankasem / @jcocchi , I've fixed all the merge conflicts in this PR and updated the templates to the latest versions of the extensions. This should be ready for review now. Thanks!

Let me know when you've rebased and I'll be ready to review

soninaren commented:
Have ported the relevant changes to this PR 
https://github.com/Azure/azure-functions-templates/pull/1477
